numpy/core/src/multiarray/descriptor.c

```diff
@@ -1651,71 +1565,186 @@ PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
+        *at = parse_dtype_from_datetime_typestr(type, len);
+        if (*at == NULL) {
+            return NPY_FAIL;
+        }
+        /* *at has byte order '=' at this point */
+        if (!PyArray_ISNBO(endian)) {
+            (*at)->byteorder = endian;
+        }
+        return NPY_SUCCEED;
+    }
+
+    /* A typecode like 'd' */
+    if (len == 1) {
+        /* Python byte string characters are unsigned */
+        check_num = (unsigned char) type[0];
+    }
+    /* A kind + size like 'f8' */
+    else {
+        char *typeend = NULL;
+        int kind;
+
+        /* Parse the integer, make sure it's the rest of the string */
+        elsize = (int)strtol(type + 1, &typeend, 10);
+        if (typeend - type == len) {
+
+            kind = type[0];
+            switch (kind) {
+                case NPY_STRINGLTR:
+                case NPY_STRINGLTR2:
+                    check_num = NPY_STRING;
+                    break;
+
+                /*
+                 * When specifying length of UNICODE
+                 * the number of characters is given to match
+                 * the STRING interface.  Each character can be
+                 * more than one byte and itemsize must be
+                 * the number of bytes.
+                 */
+                case NPY_UNICODELTR:
+                    check_num = NPY_UNICODE;
+                    elsize <<= 2;
+                    break;
+
+                case NPY_VOIDLTR:
+                    check_num = NPY_VOID;
+                    break;
+
+                default:
+                    if (elsize == 0) {
+                        check_num = NPY_NOTYPE+10;
+                    }
+                    /* Support for generic processing c8, i4, f8, etc...*/
+                    else {
+                        check_num = PyArray_TypestrConvert(elsize, kind);
+                        if (check_num == NPY_NOTYPE) {
+                            check_num += 10;
+                        }
+                        elsize = 0;
+                    }
+            }
+        }
+    }
+
     if (PyErr_Occurred()) {
         goto fail;
     }
 
-finish:
     if ((check_num == NPY_NOTYPE + 10) ||
             (*at = PyArray_DescrFromType(check_num)) == NULL) {
         PyErr_Clear();
         /* Now check to see if the object is registered in typeDict */
-        if (typeDict != NULL) {
-            PyObject *item = NULL;
+        if (typeDict == NULL) {
+            goto fail;
+        }
+        PyObject *item = NULL;
 #if defined(NPY_PY3K)
-            if (PyBytes_Check(obj)) {
-                PyObject *tmp;
-                tmp = PyUnicode_FromEncodedObject(obj, "ascii", "strict");
-                if (tmp == NULL) {
-                    goto fail;
-                }
-                item = PyDict_GetItem(typeDict, tmp);
-                Py_DECREF(tmp);
-            }
-            else {
-                item = PyDict_GetItem(typeDict, obj);
+        if (PyBytes_Check(obj)) {
```

Comment:
This check can be removed as well I think? Won't gh-15261 create a merge conflict here? (although I suppose its not a huge one).

Reply:
gh-15261 is on hold until I fix all this

Reply:
Good catch, check removed as part of the second commit in this patch. Also some other useless code elided.